### PR TITLE
return state by default and keysets in output

### PR DESCRIPTION
**Pull request by rachchan**
### What changes were proposed in this pull request?
All the methods from the `GraphPerspective` to transform the graph into a table and all the methods from `Table` interface that used to accept a lambda returning a Row or a list of them had been replaced by data frame-like versions. Now there are two methods for this:
- `select(values, ...)`: extract the values from the state and properties of each vertex to create a Row
- `explode(values, ...)`: creates a row for every tuple of all the columns refered by values zipped together

### Why are the changes needed?
These changes allow to infer headers easily so there is no need to write them in `to_df` anymore and the output formats can use the names. For instance the CsvFormat now prints the header by default. Besides, as these algorithms don't get lambdas anymore, they can be used from Python without penalty.

### Does this PR introduce any user-facing change? If yes is this documented?
Yes, scaladoc were updates and also the docs

### How was this patch tested?
All the tests pass, and the output of many of the algorithms with no test were inspected manually

### Are there any further changes required?
The storage for the rows could be improved, but it isn't worth it spending more time in that for the time being

### CODE REVIEW

File: core/src/main/scala/com/raphtory/api/analysis/algorithm/NodeListOutput.scala [hunk 1]
```diff
@@ -1,19 +1,26 @@
 package com.raphtory.api.analysis.algorithm
 
-import com.raphtory.api.analysis.table.Row
-import com.raphtory.api.analysis.table.Table
+import com.raphtory.algorithms.generic.centrality.PageRank
+import com.raphtory.api.analysis.table.{KeyPair, Row, Table}
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/api/analysis/algorithm/NodeListOutput.scala:4]: check your IntelliJ settings to apply formatting correctly

File: core/src/main/scala/com/raphtory/formats/CsvFormat.scala [hunk 2]
```diff
@@ -46,13 +46,14 @@ case class CsvFormat(delimiter: String = ",") extends Format {
...
           str
-        else
+        } else {
           "\"" + str + "\""
+        }
 
-      private def csvValue(obj: Any): String =
-        obj match {
+      private def csvValue(obj: KeyPair): String =
+        obj.value match {
```
[the remark this answers is not in the document]
Comment (ljeub-pometry) [reply core/src/main/scala/com/raphtory/formats/CsvFormat.scala:56]: We should add column headers to the table object itself so we can make sure the output is consistent and we can handle missing values

[the remark this answers is not in the document]
Comment (ljeub-pometry) [reply core/src/main/scala/com/raphtory/formats/CsvFormat.scala:56]: At least try to catch obvious problems with the table such as swapped key pairs

File: core/src/main/scala/com/raphtory/formats/JsonFormat.scala [hunk 3]
```diff
@@ -78,7 +78,7 @@ case class JsonFormat(level: JsonFormat.Level = JsonFormat.ROW) extends Format {
   }
 
   private def printRowObject(generator: JsonGenerator, serializer: ObjectWriter, row: Row): Unit =
-    serializer.writeValue(generator, row.getValues())
+    serializer.writeValue(generator, row.getValues().map(_.value))
```
[the remark this answers is not in the document]
Comment (ljeub-pometry) [reply core/src/main/scala/com/raphtory/formats/JsonFormat.scala:81]: add `row.keys()`, `row.items()` methods and rename `getValues()` to `values()`

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoReversedView.scala [hunk 4]
```diff
@@ -9,6 +9,7 @@ import scala.collection.View
 private[pojograph] class PojoReversedVertexView[T](override val vertex: PojoConcreteVertexBase[T])
         extends PojoLocalVertexViewBase(vertex) {
 
+  override def getStateSet(): List[String] = ???
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoReversedView.scala:12]: Implement this by pointing at the underlying vertex (implementation should go in `PojoLocalVertexViewBase`

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoReversedView.scala [hunk 5]
```diff
@@ -30,7 +31,7 @@ private[pojograph] class PojoReversedVertexView[T](override val vertex: PojoConc
 private[pojograph] class PojoReducedReversedVertexView(override val vertex: PojoExVertex)
         extends PojoReversedVertexView(vertex)
         with ReducedVertex {
-
+  override def getStateSet(): List[String] = ???
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoReversedView.scala:34]: delete this as it should be inherited from above

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoReversedView.scala [hunk 6]
```diff
@@ -51,7 +52,7 @@ private[pojograph] object PojoReducedReversedVertexView {
 private[pojograph] class PojoExplodedReversedVertexView(override val vertex: PojoExplodedVertex)
         extends PojoReversedVertexView[(Long, Long)](vertex)
         with ExplodedVertex {
-
+  override def getStateSet(): List[String] = ???
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoReversedView.scala:55]: delete this as it should be inherited from above

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoUndirectedView.scala [hunk 7]
```diff
@@ -9,7 +9,7 @@ import scala.collection.View
 
 private[pojograph] class PojoUndirectedVertexView[T](override val vertex: PojoConcreteVertexBase[T])
         extends PojoLocalVertexViewBase(vertex) {
-
+  override def getStateSet(): List[String] = ???
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoUndirectedView.scala:12]: delete this as it should be inherited from above

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoUndirectedView.scala [hunk 8]
```diff
@@ -97,6 +97,8 @@ private[pojograph] class PojoExplodedUndirectedVertexView(override val vertex: P
         with ExplodedVertex {
 
   override def timestamp: Long = vertex.timestamp
+
+  override def getStateSet(): List[String] = ???
```
Comment (ljeub-pometry) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoUndirectedView.scala:101]: delete this as it should be inherited from above

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]
- Fix python `to_df()` and make sure all python tests work
- Implement missing methods
- Make sure csv format catches obvious format errors
- Json format should return a dict
- Make some tickets for implementing missing functionality

Do we get rid of the `Row` object completely and make `select` work with the table directly?

File: connectors/aws/src/main/scala/com/raphtory/aws/AwsS3Sink.scala [hunk 9]
```diff
@@ -15,149 +23,164 @@ import java.nio.charset.StandardCharsets
...
+      private var partNumber                               = 1
+      private var streamLimit                              = 5000000000L
+      private var partSize: Long                           = 5 * 1024 * 1024
       private var uploadRequest: Option[UploadPartRequest] = None
-      private var streamPosition: Long = 0L
+      private var streamPosition: Long                     = 0L
 
       private var stream: ByteArrayOutputStream = new ByteArrayOutputStream()
 
+      override def allowsHeader: Boolean      = true
```
Comment (ricopinazo) [inline connectors/aws/src/main/scala/com/raphtory/aws/AwsS3Sink.scala:93]: this is the only actual change for this and the others sinks

File: core/src/main/scala/com/raphtory/api/analysis/graphview/GraphPerspective.scala [hunk 10]
```diff
@@ -234,36 +234,20 @@ trait GraphPerspective {
...
-    */
-  def select(f: (Vertex, GraphState) => Row): Table
-
-  /** Write global graph state to table (this creates a table with a single row)
-    *
-    * @param f function to extract data from graph state (runs only once)
-    */
-  def globalSelect(f: GraphState => Row): Table
+  /** Returns a table with one row per vertex and one column for each of the vertex properties and states available */
+  def selectAll(): Table
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/api/analysis/graphview/GraphPerspective.scala:238]: Hide this inside select.

File: core/src/main/scala/com/raphtory/api/analysis/graphview/GraphViewImplementation.scala [hunk 11]
```diff
@@ -63,24 +63,13 @@ final private[raphtory] case class IterateWithGraph[V <: Vertex](
...
-
-final private[raphtory] case class ExplodeSelect[V <: Vertex](f: V => IterableOnce[Row])
-        extends TabularisingGraphFunction
-
-final private[raphtory] case class ExplodeSelectWithGraph[V <: Vertex](f: (V, GraphState) => IterableOnce[Row])
+final private[raphtory] case class GlobalSelect(values: Seq[String])
         extends TabularisingGraphFunction
         with GlobalGraphFunction
+
+final private[raphtory] case object InferHeader      extends GraphFunction
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/api/analysis/graphview/GraphViewImplementation.scala:72]: This is nice!

File: core/src/main/scala/com/raphtory/api/analysis/table/TableOutput.scala [hunk 12]
```diff
@@ -5,52 +5,63 @@ import com.raphtory.api.progresstracker._
...
+        val renamedColumns = row.columns.collect {
+          case (key, value) if newNames contains key => (newNames(key), value)
+          case tuple                                 => tuple
+        }
+        new Row(renamedColumns)
+      }
+    this.copy(rows = renamedRows, header = renamedHeader)
+  }
 
-  /** Write out data and
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/api/analysis/table/TableOutput.scala:31]: Why were these comments removed

File: core/src/main/scala/com/raphtory/internals/storage/arrow/AbstractGraphLens.scala [hunk 13]
```diff
@@ -173,11 +148,15 @@ abstract class AbstractGraphLens(
...
 
-  override def executeSelect(f: (_, GraphState) => Row, graphState: GraphState)(onComplete: () => Unit): Unit = ???
-
   override def viewUndirected()(onComplete: () => Unit): Unit = ???
 
   override def viewDirected()(onComplete: () => Unit): Unit = ???
 
   override def viewReversed()(onComplete: () => Unit): Unit = ???
+
+  override def executeSelect(values: Seq[String])(onComplete: () => Unit): Unit            = ???
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/internals/storage/arrow/AbstractGraphLens.scala:157]: Do these need to be overridden?

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/edge/PojoExEdgeBase.scala [hunk 14]
```diff
@@ -33,12 +34,15 @@ private[pojograph] trait PojoExEdgeBase[T] extends ConcreteEdge[T] {
     view.sendMessage(FilteredInEdgeMessage(view.superStep + 1, dst, src))
   }
 
-  private var computationValues: Map[String, Any] =
-    Map.empty //Partial results kept between supersteps in calculation
+  //put in pojo ex entity
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/edge/PojoExEdgeBase.scala:37]: Still need to do?

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoExVertex.scala [hunk 15]
```diff
@@ -39,6 +39,7 @@ private[raphtory] class PojoExVertex(
   // state related
   def setState(key: String, value: Any): Unit =
     computationValues += ((key, value))
+  override def getStateSet(): List[String] = computationValues.keys.toList
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoExVertex.scala:42]: Probably don't want this as list in the future

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoExplodedVertex.scala [hunk 16]
```diff
@@ -43,6 +43,8 @@ private[pojograph] class PojoExplodedVertex(
 
   override def setState(key: String, value: Any): Unit = computationValues += (key -> value)
 
+  override def getStateSet(): List[String] = computationValues.keys.toList
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/entities/external/vertex/PojoExplodedVertex.scala:46]: Same as above

File: core/src/test/scala/com/raphtory/formats/JsonFormatTest.scala [hunk 17]
```diff
@@ -35,11 +38,19 @@ class JsonFormatTest extends FunSuite {
       |  "partitionID" : 13,
       |  "perspectives" : [ {
       |    "timestamp" : 100,
-      |    "rows" : [ [ "id1", 34 ], [ "id2", 24 ] ]
+      |    "rows" : [ {
```
Comment (miratepuffin) [inline core/src/test/scala/com/raphtory/formats/JsonFormatTest.scala:41]: *chef's kiss*

File: it/src/test/scala/com/raphtory/api/analysis/graphstate/AccumulatorTest.scala [hunk 18]
```diff
@@ -56,10 +56,12 @@ object CheckNodeCount extends Generic {
     CountNodes(graph)
 
   override def tabularise(graph: GraphPerspective): Table =
-    graph.globalSelect { graphState =>
-      val n: Int = graphState("nodeCount").value
-      Row(graphState.nodeCount == n)
-    }
+    graph
+      .setGlobalState { state =>
```
Comment (miratepuffin) [inline it/src/test/scala/com/raphtory/api/analysis/graphstate/AccumulatorTest.scala:60]: Only one that looks a little weird, but all others look so much more clean!

File: it/src/test/scala/com/test/raphtory/WriteValue.scala [hunk 19]
```diff
@@ -11,4 +13,10 @@ case class WriteValue() extends GenericReduction {
       .step(vertex => vertex.setState("testing", 1))
       .reducedView(PropertyMergeStrategy.sum[Int])
 
+  override def tabularise(graph: ReducedGraphPerspective): Table =
+    graph
+      .step { vertex =>
+        vertex.setState("vertexname", vertex.name())
```
Comment (miratepuffin) [inline it/src/test/scala/com/test/raphtory/WriteValue.scala:19]: will change to "name" - as a side point if we were ever to com back, calling "name" in select should just call the function on the vertex I think.

File: python/src/pyraphtory/algorithms/connectedcomponents.py [hunk 20]
```diff
@@ -22,4 +22,6 @@ def iterate(v: Vertex):
         return graph.step(step).iterate(iterate, 100, True)
 
     def tabularise(self, graph: TemporalGraph) -> Table:
-        return graph.select(lambda v: Row(v.name(), v[CC_LABEL]))
+        return graph.step(lambda v: v.set_state("name",v.name()))\
```
Comment (miratepuffin) [inline python/src/pyraphtory/algorithms/connectedcomponents.py:25]: Is this actually needed, or is just for testing?

File: python/src/pyraphtory/algorithms/pagerank.py [hunk 21]
```diff
@@ -28,11 +28,12 @@ def iterate(v: Vertex):
...
                 v.message_out_neighbours(new_label / out_degree)
-
             if abs(new_label - current_label) < 0.00001:
                 v.vote_to_halt()
 
         return graph.step(init).iterate(iterate, self.max_steps, False)
 
     def tabularise(self, graph: TemporalGraph) -> Table:
-        return graph.select(lambda v: Row(v.name(), v["prlabel"]))
+        return graph.step(lambda v: v.set_state("name",v.name()))\
```
Comment (miratepuffin) [inline python/src/pyraphtory/algorithms/pagerank.py:37]: Same as above

File: python/src/pyraphtory/algorithms/sample.py [hunk 22]
```diff
@@ -39,35 +38,39 @@ def parse(graph, tuple: str):
...
+                  .step(lambda vertex: vertex.set_state("name",vertex.name()))\
+                  .step(lambda vertex: vertex.set_state("degree",vertex.degree()))\
+                  .select("name", "degree"))\
+                  .to_df()
             print(df)
 
             df = (graph
                   .execute(PyRaphtory.algorithms.generic.community.LPA[Long]())
-                  .to_df(["name", "lpa_label"])
+                  .to_df()
```
Comment (miratepuffin) [inline python/src/pyraphtory/algorithms/sample.py:49]: 👌🏻

File: python/src/pyraphtory/algorithms/sample.py [hunk 23]
```diff
@@ -39,35 +38,39 @@ def parse(graph, tuple: str):
             graph.load(Source(lotr_spout, GraphBuilder(parse)))
 
             df = (graph
-                  .select(lambda vertex: Row(vertex.name(), vertex.degree()))
-                  .to_df(["name", "degree"]))
+                  .step(lambda vertex: vertex.set_state("name",vertex.name()))\
```
Comment (miratepuffin) [inline python/src/pyraphtory/algorithms/sample.py:41]: Just as a note incase we forget, I believe Lucas was mentioning certain phrases mapping to function calls like degree. Definitely not for now, but for it we were to come back here

File: python/src/pyraphtory/algorithms/twohoppaths.py [hunk 24]
```diff
@@ -43,13 +43,14 @@ def iterate1(v: Vertex):
...
-            return_rows = []
-            if len(r.get_values()) > 1:
-                if r.get(1) is not None:
-                    for hop_pair in r.get(1):
-                        return_rows.append(Row(r.get(0), *hop_pair))
-            else:
-                return_rows = []
-            return return_rows
-        return graph.select(lambda v: Row(v.name(), v[TWO_HOP_PATHS])).explode(explode)
+        # def explode(r: Row):
```
Comment (miratepuffin) [inline python/src/pyraphtory/algorithms/twohoppaths.py:46]: Can probably delete

File: python/src/pyraphtory/graph.py [hunk 25]
```diff
@@ -24,38 +24,25 @@ class Perspective(GenericScalaProxy):
 class Table(GenericScalaProxy):
     _classname = "com.raphtory.api.analysis.table.Table"
 
-    def to_df(self, cols):
-        rows = []
-        columns = ()
+    def to_df(self):
```
Comment (miratepuffin) [inline python/src/pyraphtory/graph.py:27]: So much cleaner

**Review by miratepuffin — APPROVED** [review]
A couple of niggly little bits, but most that we can come back to later. Smashing job @ricopinazo and @rachchan !

Going to do a couple of small fixes myself and checkout the docs before merging.

**Comment by miratepuffin** [comment]
Ok I have had a couple of issues working on this locally :D Lets quickly finish it up together tomorrow.

File: core/src/main/scala/com/raphtory/api/analysis/graphview/GraphPerspective.scala (hunk 10 above)
Comment (ricopinazo) [reply core/src/main/scala/com/raphtory/api/analysis/graphview/GraphPerspective.scala]: This is now fixed

File: core/src/main/scala/com/raphtory/api/analysis/table/TableOutput.scala (hunk 12 above)
Comment (rachchan) [reply core/src/main/scala/com/raphtory/api/analysis/table/TableOutput.scala:31]: these comments are in the superclass so no need to duplicate the docs

File: python/src/pyraphtory/algorithms/twohoppaths.py (hunk 24 above)
Comment (rachchan) [reply python/src/pyraphtory/algorithms/twohoppaths.py:46]: theres no tests for this so wasn't sure if this actually works *emoji: this is fine*

**Pull request merged**